io/directory_test.go

```diff
@@ -251,6 +254,72 @@ func TestUpgradeableDirectory(t *testing.T) {
 	compareDirectoryEntries(t, upgradedDir, dir, []*ipld.Link{missingLink})
 }
 
+// Test that we fetch as little nodes as needed to reach the HAMTShardingSize
+// during the sizeBelowThreshold computation.
+// FIXME: This only works for a sequential DAG walk.
+// FIXME: Failing in the CI for Ubuntu. This may likely be an indication of race
```

Comment:
Try rebasing this PR on the others, I can't tell whether the tests are being run with concurrent or sequential DAG walking.

With concurrent I could see how there could be a failure since walking isn't exact.

IMO the behavior we would want to be testing is that if we only have data in the shards traversed at the end that we don't just hang. However, my suspicion is that this is pretty painful to implement (@schomatis @Stebalien lmk if I'm wrong here).

A good enough approach for testing what we already have implemented is that:
1) We only get the data we need when walking serially
2) We get approximately the data we need when walking concurrently

Reply:
To some extent the amount of over-retrieval of data here depends on both the parallelism and the particulars of the parallel-walk implementation.

IMO the big thing we're trying to avoid is not so much downloading extra data, but wasting a bunch of time waiting around for that data to download unnecessarily. WDYT about if we add a 1s delay (in addition to the counting) to the blockstore for all the blocks that we do not need. I would hope that this would shrink the number of extra blocks we successfully fetch to zero.

If not then it may be that we have some bug to evaluate and decide what to do next.

Reply:
I can go in that direction if you want. My main (but not blocking) concern would be that adding a big timeout would turn the walk into a synchronous one and would defeat the purpose of the asynchronous (concurrent `dag.Walk` call) test.

> IMO the big thing we're trying to avoid is not so much downloading extra data, but wasting a bunch of time waiting around for that data to download unnecessarily.

I agree 100% on this, just not sure how to test it without a very involved (low-level, API breaking) implementation.

(Thinking out loud maybe there's not much more to gain in the asynchronous test in the sense that our current implementation doesn't really account for it but just for processing as little shards as possible without knowledge of how those shards were retrieved nor how many.)

Reply:
> My main (but not blocking) concern would be that adding a big timeout would turn the walk into a synchronous one and would defeat the purpose of the asynchronous (concurrent dag.Walk call) test.

The idea is that each block get call would take >1s (aside from the ones we need). The parallelism would mean we'd fetch the blocks we need before the blocks we don't, effectively testing that we don't wait around for blocks we don't need.

Reply:
> (aside from the ones we need)

> the blocks we need before the blocks we don't

I'm not sure I follow which are the blocks "needed". In my rough mental model any shard is acceptable to estimate the HAMT dir size, we take whichever comes first.

I'm asking this because my first interpretation of your proposal was to add a delay to _all_ blocks being fetched (for the first time, successive request get no delay because they would be cached on disk) which would mean delaying the entire traversal to leave more time for `sizeBelowThreshold` to stop it the moment it processes enough shards, but this is roughly how the current synchronous version of the test works: we delay `dag.Get` requests (by means of serializing fetches in a sync manner) so that we can stop the walk the moment `sizeBelowThreshold` has enough shards and we don't over-request.

Reply:
(If I don't get any follow-up clarifications tomorrow I'll attempt a rough implementation of what I'm interpreting here to have some concrete code to discuss around which might helps us better delineate this test.)

Reply:
Moving work and discussion to https://github.com/ipfs/go-unixfs/pull/106.

Reply:
As discussed with Adin we'll move forward with the tests in their current state and adapt them as we see fit later. The important thing is to make visible to the user how many fetches we are expecting in the concurrent walk.